{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8932953c-18a6-47eb-ac20-fe911ced2505",
   "metadata": {},
   "outputs": [],
   "source": [
    "# SeqIO is a module developed for working with fasta files and sequences. It is part of the larger package called Bio \n",
    "from Bio import SeqIO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "c6ac195a-736f-4ab6-b2de-984157d923fe",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Updated FASTA file for gene ha has been saved as data/updated/h5nx_ha.fasta.\n",
      "Updated FASTA file for gene na has been saved as data/updated/h5nx_na.fasta.\n",
      "Updated FASTA file for gene mp has been saved as data/updated/h5nx_mp.fasta.\n",
      "Updated FASTA file for gene pb1 has been saved as data/updated/h5nx_pb1.fasta.\n",
      "Updated FASTA file for gene pa has been saved as data/updated/h5nx_pa.fasta.\n",
      "Updated FASTA file for gene pb2 has been saved as data/updated/h5nx_pb2.fasta.\n",
      "Updated FASTA file for gene np has been saved as data/updated/h5nx_np.fasta.\n",
      "Updated FASTA file for gene ns has been saved as data/updated/h5nx_ns.fasta.\n"
     ]
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "\n",
    "# File paths\n",
    "InFileName = \"data/metadata-with-megaclade.tsv\"  # Metadata file with strain names and dates\n",
    "path_to_fasta_files = \"data/\"  # Directory for FASTA files\n",
    "SEGMENTS = [\"ha\", \"na\", \"mp\", \"pb1\", \"pa\", \"pb2\", \"np\", \"ns\"]\n",
    "\n",
    "# Open the metadata file\n",
    "InFile = open(InFileName, 'r')\n",
    "\n",
    "# Initialize an empty dictionary to store strain names and corresponding dates\n",
    "metadata_dict = {}\n",
    "\n",
    "# Read header and get indexes for strain and date columns\n",
    "Header = InFile.readline().strip().split('\\t')\n",
    "strain_index = Header.index(\"strain\")  # Index of the strain name in metadata\n",
    "date_index = Header.index(\"date\")  # Index of the date in metadata\n",
    "\n",
    "# Process the metadata and store the strain name and corresponding date\n",
    "for Line in InFile:\n",
    "    ElementList = Line.strip().split('\\t')  # Split the line into columns\n",
    "\n",
    "    #need it to read the date column (4th column) = [3] and replace XX with 01\n",
    "    H5RawDate = ElementList[3]\n",
    "    FixedDate = H5RawDate.replace('XX', '01')\n",
    "    \n",
    "    strain_name = ElementList[strain_index]  # Get the strain name\n",
    "    date = ElementList[date_index]  # Get the date\n",
    "    \n",
    "    # Add to dictionary\n",
    "    metadata_dict[strain_name] = date\n",
    "\n",
    "# Close the metadata file after processing\n",
    "InFile.close()\n",
    "\n",
    "# Loop through each gene segment and process the FASTA files\n",
    "for gene in SEGMENTS:\n",
    "    # Define input and output FASTA file paths\n",
    "    input_fasta = path_to_fasta_files + \"h5nx_\" + gene + \".fasta\"\n",
    "    output_fasta = path_to_fasta_files + \"updated/h5nx_\" + gene + \".fasta\"  # Output file with updated strain names\n",
    "\n",
    "    # Open the output FASTA file for writing\n",
    "    OutFile = open(output_fasta, 'w')\n",
    "\n",
    "    # Parse each sequence in the input FASTA file\n",
    "    for seq in SeqIO.parse(input_fasta, \"fasta\"):\n",
    "        strain_name = seq.id  # Extract the strain name from the FASTA header\n",
    "        \n",
    "        # Check if strain name exists in the metadata dictionary\n",
    "        if strain_name in metadata_dict:\n",
    "            # Append the corresponding date to the strain name with a pipe separator\n",
    "            updated_strain_name = f\"{strain_name}|{metadata_dict[strain_name]}\"\n",
    "            seq.id = updated_strain_name  # Update the strain name in the sequence object\n",
    "            seq.description = updated_strain_name\n",
    "        else: \n",
    "            print(f\"Strain not found in metadata: {strain_name}\")\n",
    "        # Write the updated sequence to the output file\n",
    "        SeqIO.write(seq, OutFile, \"fasta\")\n",
    "\n",
    "    # Close the output file after writing all sequences\n",
    "    OutFile.close()\n",
    "\n",
    "    print(f\"Updated FASTA file for gene {gene} has been saved as {output_fasta}.\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "37ea2831-428a-46cd-95ca-67c880dbcca9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# #This script is meant to append the corresponding date to the end of the strain name with a pipe\n",
    "# #Backbone written by Louise\n",
    "\n",
    "# #Read the metadata file with updated dates\n",
    "# #Write the date to the end of the strain name in each of the fasta files \n",
    "\n",
    "\n",
    "# #File paths\n",
    "# InFileName = \"metadata-with-megaclade-updateddates.tsv\" #MEGACLADE column included in the metadata\n",
    "# InFile = open(InFileName, 'r')\n",
    "\n",
    "# SEGMENTS = [\"ha\",\"na\",\"mp\",\"pb1\",\"pa\",\"pb2\",\"np\",\"ns\"]\n",
    "# path_to_fasta_files = \"data/\"\n",
    "\n",
    "# #at the end of this for loop I want to update/write into these fasta files the data appended to the strain name\n",
    "# for gene in SEGMENTS: \n",
    "#     # reads in a fasta file, and iterates through every sequence in it\n",
    "#     input_fasta = path_to_fasta_files + \"h5nx_\"+ gene + \".fasta\"\n",
    "#     OutFile = open(input_fasta, 'w') #opens all of the fasta files \n",
    "\n",
    "#     for Line in InFile: \n",
    "#     if LineNumber > 0:\n",
    "#         Line = Line.strip('\\n') #remove line ending characters (is this necessary?)\n",
    "#         ElementList = Line.split('\\t') #not necessary since tsv = tab separated value? (okay maybe necessary, keep it in)\n",
    "#             # seq.id = the name, seq.seq = the sequence itself \n",
    "#             for seq in SeqIO.parse(input_fasta, \"fasta\"): #tells SeqIO.parse() to expect a fasta file\n",
    "#                 sequence = str(seq.id)  #seq.id is an object - this converts it to a string\n",
    "#                 sequence = ElementList[0]\n",
    "#                 print(sequence[0])\n",
    "\n",
    "# InFile.close() \n",
    "\n",
    "    \n",
    "        # if sequence.endswith(gene): \n",
    "        #     InFile.write(seq.id + \"\\n\")\n",
    "        #     # print(seq.id, \"gene name at end of sequence\")\n",
    "        #     count += 1\n",
    "        # elif gene in sequence: \n",
    "        #     InFile.write(seq.id + \"\\n\")\n",
    "        #     # print(seq.id, \"gene within sequence\")\n",
    "        #     count +=1\n",
    "        # else: \n",
    "        #     pass\n",
    "    # print(count)\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:base] *",
   "language": "python",
   "name": "conda-base-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
