{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "2cb99149-aaf7-490c-96a4-7e95e40dea48",
   "metadata": {},
   "outputs": [],
   "source": [
    "# SeqIO is a module developed for working with fasta files and sequences. It is part of the larger package called Bio \n",
    "from Bio import SeqIO\n",
    "import re"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "b8a41a3d-1564-438a-94e9-782934efd32a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ha\n",
      "127\n",
      "na\n",
      "127\n",
      "mp\n",
      "127\n",
      "pb1\n",
      "127\n",
      "pa\n",
      "127\n",
      "pb2\n",
      "127\n",
      "np\n",
      "127\n",
      "ns\n",
      "127\n"
     ]
    }
   ],
   "source": [
    "# #This script (backbone written by Louise.. so far) identifies sequences that have the segment name added to the sequence\n",
    "# #aka identifies sequences with letters that are not atcgn \n",
    "\n",
    "# #Now I need to append these strain names to an exclude txt file\n",
    "\n",
    "# SEGMENTS = [\"ha\",\"na\",\"mp\",\"pb1\",\"pa\",\"pb2\",\"np\",\"ns\"]\n",
    "\n",
    "# path_to_fasta_files = \"data/\"\n",
    "\n",
    "# #File paths\n",
    "# InFileName = \"config/exclude_strains.txt\"\n",
    "# InFile = open(InFileName, 'a')\n",
    "\n",
    "# # Use a set to collect unique sequence IDs\n",
    "# unique_ids = set()\n",
    "\n",
    "# # reads in a fasta file, and iterates through every sequence in it\n",
    "# for gene in SEGMENTS: \n",
    "#     count = 0\n",
    "#     print(gene)\n",
    "#     input_fasta = path_to_fasta_files + \"h5nx_\"+ gene + \".fasta\"\n",
    "     \n",
    "#     # seq.id = the name \n",
    "#     # seq.seq = the sequence itself \n",
    "#     for seq in SeqIO.parse(input_fasta, \"fasta\"): \n",
    "#         sequence = str(seq.seq)  #seq.seq is an object - this converts it to a string \n",
    "#         sequence_id = str(seq.id) #seq.id is an object - this converts it to a string\n",
    "#         #     print(seq.seq)\n",
    "#         if sequence.endswith(gene): \n",
    "#             unique_ids.add(seq.id + \"\\n\")\n",
    "#             # print(seq.id, \"gene name at end of sequence\")\n",
    "#             # count += 1\n",
    "#         elif gene in sequence: \n",
    "#             unique_ids.add(seq.id + \"\\n\")\n",
    "#             # print(seq.id, \"gene within sequence\")\n",
    "#             # count +=1\n",
    "#         elif re.match(r'^r\\d+', sequence_id): # r before the string tells Python to treat the string as \"raw\", meaning it doesn't interpret escape sequences (like \\d)\n",
    "#             unique_ids.add(sequence_id + \"\\n\") \n",
    "#             # print(sequence_id, \"starts with r\")\n",
    "#             count += 1\n",
    "#         else: \n",
    "#             pass\n",
    "#     print(count)\n",
    "\n",
    "# # Append the unique sequence IDs to the output file\n",
    "# with open(InFileName, 'a') as InFile:  # Open file in append mode to keep previous data\n",
    "#     for seq_id in unique_ids:\n",
    "#         InFile.write(seq_id + \"\\n\")\n",
    "\n",
    "# InFile.close() \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "4b05dca7-9d8d-4970-9215-729bdf4954f1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ha\n",
      "Total sequences for ha: 174\n",
      "na\n",
      "Total sequences for na: 717\n",
      "mp\n",
      "Total sequences for mp: 175\n",
      "pb1\n",
      "Total sequences for pb1: 127\n",
      "pa\n",
      "Total sequences for pa: 174\n",
      "pb2\n",
      "Total sequences for pb2: 127\n",
      "np\n",
      "Total sequences for np: 173\n",
      "ns\n",
      "Total sequences for ns: 174\n",
      "Total unique sequences appended to config/exclude_strains.txt: 719\n"
     ]
    }
   ],
   "source": [
    "import re\n",
    "from Bio import SeqIO\n",
    "\n",
    "SEGMENTS = [\"ha\", \"na\", \"mp\", \"pb1\", \"pa\", \"pb2\", \"np\", \"ns\"]\n",
    "\n",
    "path_to_fasta_files = \"data/\"\n",
    "\n",
    "# File paths\n",
    "InFileName = \"config/exclude_strains.txt\"\n",
    "\n",
    "# Use a set to collect unique sequence IDs\n",
    "unique_ids = set()\n",
    "\n",
    "# Reads in a fasta file, and iterates through every sequence in it\n",
    "for gene in SEGMENTS: \n",
    "    count = 0\n",
    "    print(gene)\n",
    "    input_fasta = path_to_fasta_files + \"h5nx_\" + gene + \".fasta\"\n",
    "     \n",
    "    # seq.id = the name\n",
    "    # seq.seq = the sequence itself \n",
    "    for seq in SeqIO.parse(input_fasta, \"fasta\"): \n",
    "        sequence = str(seq.seq)  # seq.seq is an object - this converts it to a string \n",
    "        sequence_id = str(seq.id) # seq.id is an object - this converts it to a string\n",
    "\n",
    "        # Check if the sequence ends with the gene name\n",
    "        if sequence.endswith(gene): \n",
    "            unique_ids.add(seq.id)  # Add the sequence ID to the set\n",
    "            count += 1  # Increment count for this condition\n",
    "\n",
    "        # Check if the gene is present within the sequence\n",
    "        elif gene in sequence: \n",
    "            unique_ids.add(seq.id)  # Add the sequence ID to the set\n",
    "            count += 1  # Increment count for this condition\n",
    "\n",
    "        # Check if the sequence ID starts with 'r' followed by digits, OK but this doesn't exclude all of the Laboratoryderived seqs\n",
    "        elif re.match(r'^r\\d+', sequence_id):  # Using regular expression\n",
    "            unique_ids.add(sequence_id)  # Add the sequence ID to the set\n",
    "            count += 1  # Increment count for this condition\n",
    "\n",
    "        else: \n",
    "            pass\n",
    "\n",
    "    print(f\"Total sequences for {gene}: {count}\")\n",
    "\n",
    "# Append the unique sequence IDs to the output file\n",
    "with open(InFileName, 'a') as InFile:  # Open file in append mode to keep previous data\n",
    "    for seq_id in unique_ids:\n",
    "        InFile.write(seq_id + \"\\n\")\n",
    "\n",
    "print(f\"Total unique sequences appended to {InFileName}: {len(unique_ids)}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "f8bd23df-89e6-4865-a5dc-df38962460b2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ha\n",
      "Total sequences for ha: 174\n",
      "na\n",
      "Total sequences for na: 717\n",
      "mp\n",
      "Total sequences for mp: 175\n",
      "pb1\n",
      "Total sequences for pb1: 127\n",
      "pa\n",
      "Total sequences for pa: 174\n",
      "pb2\n",
      "Total sequences for pb2: 127\n",
      "np\n",
      "Total sequences for np: 173\n",
      "ns\n",
      "Total sequences for ns: 174\n",
      "Total unique sequences appended to config/exclude_strains.txt: 719\n"
     ]
    }
   ],
   "source": [
    "import re\n",
    "from Bio import SeqIO\n",
    "\n",
    "SEGMENTS = [\"ha\", \"na\", \"mp\", \"pb1\", \"pa\", \"pb2\", \"np\", \"ns\"]\n",
    "\n",
    "path_to_fasta_files = \"data/\"\n",
    "\n",
    "# File paths\n",
    "InFileName = \"config/exclude_strains.txt\"\n",
    "\n",
    "# Use a set to collect unique sequence IDs\n",
    "unique_ids = set()\n",
    "\n",
    "# Reads in a fasta file, and iterates through every sequence in it\n",
    "for gene in SEGMENTS: \n",
    "    count = 0\n",
    "    print(gene)\n",
    "    input_fasta = path_to_fasta_files + \"h5nx_\" + gene + \".fasta\"\n",
    "     \n",
    "    # seq.id = the name\n",
    "    # seq.seq = the sequence itself \n",
    "    for seq in SeqIO.parse(input_fasta, \"fasta\"): \n",
    "        sequence = str(seq.seq)  # seq.seq is an object - this converts it to a string \n",
    "        sequence_id = str(seq.id) # seq.id is an object - this converts it to a string\n",
    "\n",
    "        # Check if the sequence ends with the gene name\n",
    "        if sequence.endswith(gene): \n",
    "            unique_ids.add(seq.id)  # Add the sequence ID to the set\n",
    "            count += 1  # Increment count for this condition\n",
    "\n",
    "        # Check if the gene is present within the sequence\n",
    "        elif gene in sequence: \n",
    "            unique_ids.add(seq.id)  # Add the sequence ID to the set\n",
    "            count += 1  # Increment count for this condition\n",
    "\n",
    "        # Check if the sequence ID starts with 'r' followed by digits, OK but this doesn't exclude all of the Laboratoryderived seqs\n",
    "        elif re.match(r'^r\\d+', sequence_id):  # Using regular expression\n",
    "            unique_ids.add(sequence_id)  # Add the sequence ID to the set\n",
    "            count += 1  # Increment count for this condition\n",
    "\n",
    "        else: \n",
    "            pass\n",
    "\n",
    "    print(f\"Total sequences for {gene}: {count}\")\n",
    "\n",
    "# Append the unique sequence IDs to the output file\n",
    "with open(InFileName, 'a') as InFile:  # Open file in append mode to keep previous data\n",
    "    for seq_id in unique_ids:\n",
    "        InFile.write(seq_id + \"\\n\")\n",
    "\n",
    "print(f\"Total unique sequences appended to {InFileName}: {len(unique_ids)}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5a8e4653-2cb7-4497-891c-ac9fe40d3b34",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:base] *",
   "language": "python",
   "name": "conda-base-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
